Sources/SecureXPC/Server/XPCServer.swift

```diff
@@ -177,6 +177,46 @@ public class XPCServer {
     // Routes
     private var routes = [XPCRoute : XPCHandler]()
     
+    /// Set of weak references to connections, used to update their dispatch queues.
+    private var connections = Set<WeakConnection>()
+    
+    /// Weak wrapper around a connection stored in the `connections` variable.
+    ///
+    /// Designed to be conveniently settable as the context of an `xpc_connection_t` so that it's accessible from its finalizer.
+    fileprivate class WeakConnection: Hashable {
+        private weak var server: XPCServer?
+        fileprivate weak var connection: xpc_connection_t?
+        private let connectionHash: Int
+        
+        init(_ connection: xpc_connection_t, server: XPCServer) {
+            self.connection = connection
+            self.connectionHash = xpc_hash(connection)
+            self.server = server
+        }
+        
+        func removeFromContainer() {
+            self.server?.connections.remove(self)
+        }
+        
+        static func == (lhs: XPCServer.WeakConnection, rhs: XPCServer.WeakConnection) -> Bool {
+            lhs.connectionHash == rhs.connectionHash
+        }
```

Comment:
Why not delegate to `xpc_equal`?

Reply:
Because `xpc_equal` doesn't accept `nil` values so it would've been a bunch more lines of code to write, but good point about the possibilities of hash collisions. Will fix.

Reply:
I see two approaches:

```suggestion
        static func == (lhs: XPCServer.WeakConnection, rhs: XPCServer.WeakConnection) -> Bool {
            xpc_equal(lhs.connection ?? xpc_null_create(), rhs.connection ?? xpc_null_create())
        }
```

```suggestion
        static func == (lhs: XPCServer.WeakConnection, rhs: XPCServer.WeakConnection) -> Bool {
            switch (lhs.connection, rhs.connection) {
                case (nil, nil), (nil, _?), (_?, nil): return false
                case (let a, let b): return xpc_equal(a, b)
            }
        }
```

Reply:
Nevermind on that second approach, I realized that despite one being `nil` compared to another, the values might still need to be considered equal (in order for removal from the array to work)

Reply:
Wasn't aware switch statements could be used like that. I pushed a version using if/else if/else, want to take a look?

Reply:
Actually, I think this approach is general (of trying to do equality "after death") is irreparably flawed.

The hashes can't be relied on for inequality because they're insufficient. The insufficiency causes false positives. The false positives can cause false matches against a (different) item in the array. That then fucks everything way up.

I think the connections array needs to be reworked into a dictionary.

Given that there's no "xpc_id()` (from what I've seen) where `(xpc_id(a) == xpc_id(b)) ≣ xpc_equal(a, b)`, I think we would need to use a `UUID`, counter (backed by an atomic int or whatever),, or some other durable identifier we generate for ourselves.

Reply:
Swift has really awesome pattern matching. Almost all `if`/`else if`/`else` ladders are more customarily be expressed with a `switch`. https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html#ID127

Though there are some exceptions (e.g. somewhere in the code base there's a ladder that calls `xpc_equal` a bunch of times. There's no good way to convert that to a `switch`)

Reply:
> I think the connections array needs to be reworked into a dictionary.

> Given that there's nothing like an xpc_id() function (from what I've seen) where (xpc_id(a) == xpc_id(b)) ≣ xpc_equal(a, b), I think we would need to use a UUID, counter (backed by an atomic int or whatever), or some other durable identifier we generate for ourselves as keys to the dict. These would be durable and exact, and will work even after the connection dies.

It's currently a set, not an array. I think a set can continue to work, but the unique identification can be UUID based instead of hashing the connection. That'd suffice, yeah?

Reply:
Yeah, that can work, if you just implement `Hashable` and `Equatable` solely in terms of that `UUID` (and perhaps makes sense to lump in `Identifiable`, too).

Though IMO a dictionary is more fitting. You just have `[UUID: Weak<Connection>]`. You would no longer need conformances to `Hashable` and `Equatable`.

Reply:
Went with the `UUID` approach and kept it a set. Using a set felt a bit clearer to me as it's a less leaky abstraction where the rest of the class when interacting with the connections doesn't need to know how to add or iterate over them.